File: mpmt1.erl
%
%  mpmt1.erl: A stupid simple example of Erlang threading
%
% License:
%   Apache License, Version 2.0
% History:
%   * 2024/05/18 v0.1 Initial version
% Author:
%   Masanori Itoh <masanori.itoh@gmail.com>
% Usage:
%  $ erlc mpmt1.erl
%  $ erl -noshell -pa mpmt1.beam  -s mpmt1 start NUM_CTX DURATION -s init stop
%  NUM_CTX : number of threads
%  DURATION : duration in seconds
% TODO:
%   * Enable to flush remaining messages in busy_worker()
%   * Use argparse
%   * Use message to synchronize waiting for thread termination
%
-module(mpmt1).
-export([start/1, busy_worker/3, create_busy_worker/2]).

busy_worker(Idx, Current_ns, Timeleft_ns) ->
    if
        Timeleft_ns =< 0 -> 
            io:format("Expired: ~p ~p ~n", [Idx, Timeleft_ns]),
            %exit(normal);
            ok;
        true -> 
            Now = erlang:system_time(),
            busy_worker(Idx, Now, Now - Current_ns)
    end.

create_busy_worker(0, _) ->
    ok;
create_busy_worker(Num_context, Duration) ->
    io:format("create_busy_worker(): Idx: ~p~n", [Num_context]),
    % system_time() returns the timestamp in nano-seconds
    Now = erlang:system_time(),
    spawn(mpmt1, busy_worker, [Num_context, Now, Duration * 1000 * 1000 * 1000]),
    create_busy_worker(Num_context - 1, Duration).

start(Args) ->
    Duration =
        if
            length(Args) >=2  ->
                erlang:list_to_integer(atom_to_list(lists:nth(2, Args)));
            true ->
                5
        end,

    Num_context =
        if
            length(Args) >=1  ->
                erlang:list_to_integer(atom_to_list(lists:nth(1, Args)));
            true ->
                4
        end,
    
    io:format("NUM_CONTEXT: ~p DURATION: ~p~n", [Num_context, Duration]),

    create_busy_worker(Num_context, Duration),

    % TODO(thatsdone): Use message to synchronize with threads.
    io:format("start(): sleeping ~p sec.~n", [Duration + 1]),
    timer:sleep((Duration + 1) * 1000).

File: mpmt1.ml
(*
 * mpmt1.ml: A simple CPU load generator example using OCaml threading
 *
 *  License:
 *    Apache License, Version 2.0
 * History:
 *   2024/06/23 v0.1 Initial version
 * Author:
 *   Masanori Itoh <masanori.itoh@gmail.com>
 * Usage:
 *   - $ docker run --rm -it -v `pwd`:/work  ocaml/opam:debian-11-ocaml-5.3 /bin/bash
 *   - $ ocaml -I +unix -I +threads mpmt1.ml [-n NUM_CONTEXT] [-d DURATION]
 * TODO:
 * - ...
 *)

#load "unix.cma";;
#load "threads.cma";;

let timestamp () = (Float.mul (Unix.gettimeofday()) 1000.)

let rec busy_worker (current, time_left) =
   (* Printf.printf "current: %f time_left %f\n" current time_left; *)
   if time_left > 0. then
     let now = timestamp() in
     let diff = Float.sub now current in
     (* Printf.printf "now: %f diff: %f\n" now diff; *)
     busy_worker (now, (Float.sub time_left diff))
   else
     Printf.printf "Expired.\n"

let num_context = ref 3;;
let duration = ref 5;;
Arg.parse [
  ("-n", Arg.Int (function i -> num_context := i), "");
  ("-d", Arg.Int (function i -> duration := i), "");
] (function s -> ()) "ERROR";;
     
Printf.printf "num_context: %d duration: %d\n" !num_context !duration;;

let t = timestamp();;

(*
busy_worker (t,  (Float.mul (float_of_int !duration) 1000.));;
0.));;
*)

for i = 1 to !num_context do
  let _ = Thread.create busy_worker (t, (Float.mul (float_of_int !duration) 1000.)) in ()
done;;
(*
 TODO(thatsdone): Hold Thread.create() result and use Thread.join()
*)
Printf.printf "DEBUG: Sleeping... %d seconds\n" !duration;;
Unix.sleep(!duration);;

